File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.py
class Solution:
    def smallestDivisor(self, nums: List[int], k: int) -> int:
        low=1
        high=max(nums)
        while(low<=high):
            mid=(low+high)//2
            sum=0
            for i in nums:
                sum+=(i+mid-1)//mid
            if(sum<=k):
                res=mid
                high=mid-1
            else:
                low=mid+1
        return res

File: 1283-find-the-smallest-divisor-given-a-threshold/1283-find-the-smallest-divisor-given-a-threshold.java
class Solution {
    
    int m(int[] arr){
        int ma=arr[0];
        for(int i=0;i<arr.length;i++){
            if(arr[i]>ma) ma=arr[i];
        }
        return ma;
    }
    public int smallestDivisor(int[] nums, int threshold) {
        int low=1;
        int high=m(nums);
        int ans=-1;
        while (low<=high){
            int sum=0;
            int mid=low+(high-low)/2;
            for (int el:nums){
                sum+=(el+mid-1)/mid;
            }
            if (sum<=threshold){
            ans=mid;
            high=mid-1;
            }
            else low=mid+1;

        }
        return ans;
    }
}